File: baekjoonn/14888_연산자 끼워넣기/B_14888(시간초과).js
const permutations = function(idx, length) {
    if (idx === length) {
        ops.push(temp.slice())
    } else {
        for (let i = 0; i < length; i++) {
            if (!check[i]) {
                check[i] = !check[i]
                temp.push(operations[i])
                permutations(idx + 1, length)
                check[i] = !check[i]
                temp.pop()
            }
        }
    }
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const N = parseInt(input[0])
const nums = input[1].split(" ").map(v => parseInt(v))
const o = input[2].split(" ").map(v => parseInt(v))

let operations = []
for (let i = 0; i < 4; i++) {
    for (let j = 0; j < o[i]; j++) {
        if (i === 0) {
            operations.push("+")
        } else if (i === 1) {
            operations.push("-")
        } else if (i === 2) {
            operations.push("*")
        } else {
            operations.push("/")
        }
    }
}
const operationsLength = operations.length

let ops = []
let temp = []
let check = new Array(operationsLength).fill(0)
permutations(0, operationsLength)

let maxSums = -Infinity
let minSums = Infinity
for (let i = 0; i < ops.length; i++) {
    let sums = nums[0]
    let arr = ops[i]
    for (let j = 0; j < arr.length; j++) {
        if (arr[j] === '+') {
            sums = sums + nums[j + 1]
        } else if (arr[j] === '-') {
            sums = sums - nums[j + 1]
        } else if (arr[j] === '*') {
            sums = sums * nums[j + 1]
        } else {
            if (sums < 0) {
                sums = sums * -1
                sums = Math.floor(sums / nums[j + 1])
                sums = sums * -1
            } else {
                sums = Math.floor(sums / nums[j + 1])
            }
        }
    }
    if (sums > maxSums) {
        maxSums = sums
    } else if (sums < minSums) {
        minSums = sums
    }
}

console.log(maxSums)
console.log(minSums)

File: baekjoon/14888_연산자 끼워넣기/B_14888(시간초과).py
import math


def permutation(idx):
    if idx == N - 1:
        operator_perm.append(perm[:])
        return
    else:
        for k in range(N - 1):
            if not check[k]:
                check[k] = 1
                perm.append(operators[k])
                permutation(idx + 1)
                check[k] = 0
                perm.pop()


N = int(input())
nums = list(map(int, input().split()))
operator = list(map(int, input().split()))

operators = []
for i in range(len(operator)):
    if operator[i] != 0:
        for j in range(operator[i]):
            if i == 0:
                operators.append("+")
            elif i == 1:
                operators.append("-")
            elif i == 2:
                operators.append("*")
            else:
                operators.append("//")

perm = []
# 연산자 순열
operator_perm = []
check = [0] * (N - 1)
# 순열 함수
permutation(0)

# 연산자 순열에서 하나씩 꺼내서 값 계산
max_sums = -float('inf')
min_sums = float('inf')

for operate in operator_perm:
    sums = nums[0]
    for o in range(len(operate)):
        if operate[o] == "+":
            sums += nums[o + 1]
        elif operate[o] == "-":
            sums -= nums[o + 1]
        elif operate[o] == "*":
            sums *= nums[o + 1]
        else:
            if sums < 0:
                division = math.ceil(sums / nums[o + 1])
                sums = division
            else:
                sums //= nums[o + 1]
    if max_sums < sums:
        max_sums = sums
    if min_sums > sums:
        min_sums = sums

print(max_sums)
print(min_sums)
